pkg/adaptor/cloud/cloud.go

```diff
@@ -0,0 +1,420 @@
+
+	return &pb.CreateVMResponse{AgentSocketPath: socketPath}, nil
+}
+
+func (s *cloudService) StartVM(ctx context.Context, req *pb.StartVMRequest) (res *pb.StartVMResponse, err error) {
+
+	defer func() {
+		if err != nil {
+			logger.Print(err)
+		}
+	}()
+
+	sid := sandboxID(req.Id)
+
+	sandbox, err := s.getSandbox(sid)
+	if err != nil {
+		return nil, err
+	}
+
+	instance, err := s.provider.CreateInstance(ctx, sandbox.pod, string(sid), sandbox.cloudConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create an instance : %w", err)
+	}
+
+	if err := s.setInstance(sid, instance.ID, instance.Name); err != nil {
+		return nil, err
+	}
+
+	logger.Printf("created an instance %s for sandbox %s", instance.Name, sid)
+
+	if err := s.workerNode.Setup(sandbox.netNSPath, instance.IPs, sandbox.podNetworkConfig); err != nil {
+		return nil, fmt.Errorf("failed to set up pod network tunnel on netns %s: %w", sandbox.netNSPath, err)
+	}
+
+	serverURL := &url.URL{
+		Scheme: "http",
+		Host:   net.JoinHostPort(instance.IPs[0].String(), s.daemonPort),
+		Path:   forwarder.AgentURLPath,
+	}
+
+	errCh := make(chan error)
+	go func() {
+		defer close(errCh)
+
+		if err := sandbox.agentProxy.Start(context.Background(), serverURL); err != nil {
+			logger.Printf("error running agent proxy: %v", err)
+			errCh <- err
+		}
+	}()
+
+	select {
+	case <-ctx.Done():
+		_ = sandbox.agentProxy.Shutdown()
+		return nil, ctx.Err()
+	case err := <-errCh:
+		return nil, err
+	case <-sandbox.agentProxy.Ready():
+	}
+
+	logger.Printf("agent proxy is ready")
+	return &pb.StartVMResponse{}, nil
+}
+
+func (s *cloudService) StopVM(ctx context.Context, req *pb.StopVMRequest) (*pb.StopVMResponse, error) {
+
+	sid := sandboxID(req.Id)
+
+	sandbox, err := s.getSandbox(sid)
+	if err != nil {
+		err = fmt.Errorf("failed to stop VM: %v", err)
+		logger.Print(err)
+		return nil, err
+	}
+
+	if err := sandbox.agentProxy.Shutdown(); err != nil {
+		logger.Printf("failed to stop agent proxy: %v", err)
+	}
+
+	if err := s.provider.DeleteInstance(ctx, sandbox.instanceID); err != nil {
+		logger.Printf("failed to delete instance %s: %v", sandbox.instanceID, err)
+	}
+
+	if err := s.workerNode.Teardown(sandbox.netNSPath, sandbox.podNetworkConfig); err != nil {
+		logger.Printf("failed to tear down netns %s: %v", sandbox.netNSPath, err)
+	}
+
+	if err = s.removeSandbox(sid); err != nil {
+		logger.Printf("failed to remove sandbox %s: %v", sid, err)
+	}
+
+	return &pb.StopVMResponse{}, nil
+}
+
+func getPodName(annotations map[string]string) string {
+
+	sandboxName := annotations[cri.SandboxName]
+
+	// cri-o stores the sandbox name in the form of k8s_<pod name>_<namespace>_<uid>_0
+	// Extract the pod name from it.
+	if tmp := strings.Split(sandboxName, "_"); len(tmp) > 1 && tmp[0] == "k8s" {
+		return tmp[1]
+	}
+
+	return sandboxName
+}
+
+func getPodNamespace(annotations map[string]string) string {
```

Comment:
would it make sense to have these helper methods under util ?

Reply:
This function is private, and only used in this package. Should we move this function to the util package?

Reply:
I was thinking if we move this function to util and make it public it'll be useful in general -e.g for csi wrapper or other stuffs. 
Let me know what you think?

Reply:
OK. I move the function to the `util` package.